test/e2e/upgrade_management_cluster_test.go

```diff
@@ -35,307 +24,128 @@ func upgradeManagementCluster() {
-			// pass reference to the management cluster hosting this test
-			KubeconfigPath: upgradeClusterProxy.GetKubeconfigPath(),
-			// pass the clusterctl config file that points to the local provider repository created for this test
-			ClusterctlConfigPath: clusterctlConfigPath,
-
-			// define template variables
-			Namespace:                namespace,
-			ClusterName:              workloadClusterName,
-			Flavor:                   "upgrade-workload",
-			KubernetesVersion:        kubernetesVersion,
-			ControlPlaneMachineCount: controlPlaneMachineCount,
-			WorkerMachineCount:       workerMachineCount,
-			InfrastructureProvider:   clusterctl.DefaultInfrastructureProvider,
-			// setup clusterctl logs folder
-			LogFolder: filepath.Join(artifactFolder, "clusters", upgradeClusterProxy.GetName()),
-		})
-		Expect(workloadClusterTemplate).ToNot(BeNil(), "Failed to get the cluster template")
-
-		Logf("Applying the cluster template yaml to the cluster")
-		Expect(upgradeClusterProxy.Apply(ctx, workloadClusterTemplate)).To(Succeed())
-
-		By("Waiting for the machines to be running")
-		// Note: We cannot use waitForNumMachinesInState because these are v1alpha4 Machines!
-		Eventually(func(g Gomega) {
-			n := 0
-			machineList := &clusterv1alpha4.MachineList{}
-			opts := []client.ListOption{
-				client.InNamespace(namespace),
-				client.MatchingLabels{clusterv1.ClusterLabelName: workloadClusterName},
+	// Create the BMHs needed in the hosting namespace.
+	createBMH := func(clusterProxy framework.ClusterProxy, clusterNamespace string, clusterName string) {
+		installSplitYAML()
+		splitFiles := splitYAMLFile("bmhosts_crs.yaml", workDir)
+
+		isBootstrapProxy := !strings.HasPrefix(clusterProxy.GetName(), "clusterctl-upgrade")
+		if isBootstrapProxy {
+			// remove existing bmh from source and apply first 2 in target
+			Logf("remove existing bmh from source")
+			cmd := exec.Command("bash", "-c", "kubectl delete -f bmhosts_crs.yaml  -n metal3")
+			cmd.Dir = workDir
+			output, err := cmd.CombinedOutput()
+			Logf("Remove existing bmhs:\n %v", string(output))
+			Expect(err).To(BeNil())
+
+			// Apply secrets and bmhs for [node_0 and node_1] in the management cluster to host the target management cluster
+			for i := 0; i < 4; i++ {
+				resource, err := os.ReadFile(filepath.Join(workDir, splitFiles[i]))
+				Expect(err).ShouldNot(HaveOccurred())
+				Expect(clusterProxy.Apply(ctx, resource, []string{"-n", clusterNamespace}...)).ShouldNot(HaveOccurred())
 			}
-			g.Expect(upgradeClusterClient.List(ctx, machineList, opts...)).To(Succeed())
-			for _, machine := range machineList.Items {
-				if machine.Status.GetTypedPhase() == clusterv1alpha4.MachinePhaseRunning {
-					n++
-				}
+		} else {
+			// Apply secrets and bmhs for [node_2, node_3 and node_4] in the management cluster to host workload cluster
+			for i := 4; i < 10; i++ {
```

Comment:
how this is iterating through if you have only two/there BMHs here/below?

Reply:
we know that file are ordered `secret_node_0.yaml, bmh_node_0.yaml, ...,secret_node_n.yaml, bmh_node_n.yaml..` so the first iteration is to apply the secret then the second to apply bmh for each node